src/include/storage/checkpoint_io.h

```diff
@@ -0,0 +1,500 @@
+#include "storage/projected_columns.h"
+#include "storage/projected_row.h"
+#include "storage/write_ahead_log/log_io.h"
+
+namespace terrier::storage {
+// TODO(Zhaozhe): Should fetch block_size. Use magic number first.
+// The block size to output to disk every time
+#define CHECKPOINT_BLOCK_SIZE 4096
+// TODO(Yuning): Should allow setting number of pre-allocated buffers.
+// Use magic number before we have a setting manager.
+#define CHECKPOINT_BUF_NUM 10
+
+// checkpoint file header flags
+#define P_IS_CATALOG 0
+
+/**
+ * The header of a page in the checkpoint file.
+ */
+// TODO(Zhaozhe, Mengyang): More fields can be added to header
+class PACKED CheckpointFilePage {
+ public:
+  /**
+   * Initialize a given page
+   * @param page to be initialized.
+   */
+  static void Initialize(CheckpointFilePage *page) {
+    // TODO(mengyang): support non-trivial initialization
+    page->checksum_ = 0;
+    page->table_oid_ = 0;
+    page->version_ = 0;
+    page->flags_ = 0;
+  }
+
+  /**
+   * Get the checksum of the page.
+   * @return checksum of the page.
+   */
+  uint32_t GetChecksum() { return checksum_; }
+
+  /**
+   * Set the checksum of the page.
+   * @param checksum
+   */
+  void SetCheckSum(uint32_t checksum) { checksum_ = checksum; }
+
+  /**
+   * Get the oid of the table whose rows are stored in this page.
+   * @return oid of the table.
+   */
+  catalog::table_oid_t GetTableOid() { return catalog::table_oid_t(table_oid_); }
+
+  /**
+   * Set the oid of the table whose rows are stored in this page.
+   * @param oid of the table.
+   */
+  void SetTableOid(catalog::table_oid_t oid) { table_oid_ = !oid; }
+
+  /**
+   * Get the pointer to the first row in this page.
+   * @return pointer to the first row.
+   */
+  byte *GetPayload() { return payload_; }
+
+  /**
+   * Set all the flags at the same time.
+   * @param flags to be set.
+   */
+  void SetFlags(uint32_t flags) { flags_ = flags; }
+
+  /**
+   * Set the bit in the flags to some value.
+   * @param pos position of the flag to set, from the most significant bit to the least significant bit.
+   * @param value bool value to set
+   */
+  void SetFlag(uint32_t pos, bool value) { Bitmap().Set(pos, value); }
+
+  /**
+   * Get the status of the flag at some position
+   * @param pos position of the flag to test, from the most significant bit the the least significant bit.
+   * @return the value of the flag.
+   */
+  bool GetFlag(uint32_t pos) { return Bitmap().Test(pos); }
+
+  void SetVersion(uint32_t version) { version_ = version; }
+
+  uint32_t GetVersion() { return version_; }
+
+ private:
+  uint32_t checksum_;
+  // Note: use uint32_t instead of table_oid_t, because table_oid_t is not POD so that
+  // we cannot use PACKED in this class.
+  uint32_t table_oid_;
+  uint32_t version_;  // version of the schema(useful with schema change), or, for catalog tables,
+                      // serves as the type of the catalog table.
```

Comment:
This is not an optimization, but to put all the information needed in the header. There are mainly 2 reasons I didn't overload `flags_` field. 
1. Currently, there is only one bit in the `flags_` used, but maybe in the future, there will be more flags needed.
2. It may be more semantically correct to overload the table schema version field as the catalog table type, assuming catalog schemas are never changed.